base/multidimensional.jl

```diff
@@ -970,7 +970,7 @@ function diff(a::AbstractArray{T,N}; dims::Integer) where {T,N}
 end
 function diff(r::AbstractRange{T}; dims::Integer=1) where {T}
     dims == 1 || throw(ArgumentError("dimension $dims out of range (1:1)"))
```

Comment:
This should probably `require_one_based_indexing` to match other method. 

And I'm less sure, but perhaps it should apply to all AbstractVectors? (Will this increase or decrease the chance that someone thinks it ought to be `fill(step, n-1)` in the future?)

```suggestion
function diff(r::AbstractVector{T}; dims::Integer=1) where {T}
    require_one_based_indexing(r)
    dims == 1 || throw(ArgumentError("dimension $dims out of range (1:1)"))
```

Reply:
Since the implementation uses `firstindex(r):lastindex(r)-1`, it shouldn't require 1-based indexing.

Reply:
that's not relevant now I guess because we've decided to stick with `-`

Reply:
I forgot why I said this, but then I remembered. It won't read out of bounds with offset axes, but it will produce a 1-based array, while many methods on offset arrays & ranges try to preserve their indices. Whether that's desirable is the question.
```
julia> axes(OffsetArray(rand(1:10, 5), -3),1)
OffsetArrays.IdOffsetRange(values=-2:2, indices=-2:2)

julia> diff(ans)
4-element Vector{Int64}:
 1
 1
 1
 1
```

Reply:
I don't think `diff` needs to preserve because each element in "diff" was the "gap" between elements so I don't feel like indexing is super well defined anyway